### PR TITLE
Infinispan Pipeline

### PR DESCRIPTION
**Pull request by ryanemerson**
Still a WIP but most of the infinispan tests are passing.

There's a lot of changes here, but for the most part I haven't changed any controller logic. The key addition to the PR is the pipeline framework, with the existing controller logic ported over to the various Handler functions and updated to use the pipeline Context.

The most important files to review are:

```
controllers/infinispan.go
pkg/reconcile/pipeline/infinispan/api.go
pkg/reconcile/pipeline/infinispan/pipeline/pipeline.go
pkg/reconcile/pipeline/infinispan/context/resources.go
```

TODO:
- [x] Cache Service
- [x] XSite support
- [x] Hot Rod Rolling Upgrades
- [x] Remove old controller code
- [x] Test on Openshift

### CODE REVIEW

**Review by rigazilla — COMMENTED** [review]
@ryanemerson, just few comments from a quick read.
The idea of collapsing all the controllers acting on the Infinispan CR looks great to me: flow of control is clearer and we probably avoid conflicts beetween concurrent executions.

I'm going to run the testsuite locally as next step

File: pkg/reconcile/pipeline/infinispan/api.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Pipeline for Infinispan reconciliation
+type Pipeline interface {
+	// Process the pipeline
+	// Returns true if processing should be repeated and optional error if occurred
+	// important: even if error occurred it might not be needed to retry processing
+	Process(ctx context.Context) (bool, time.Duration, error)
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply pkg/reconcile/pipeline/infinispan/api.go:28]: There's definitely some overlap here. The idea of having the `Pipeline` interface as an abstraction, was to facilitate testing. The combination of this interface and the pipeline Builder mean that it's possible for us to write env/unit tests in the future that utilise the pipeline but with a testing `ContextProvider` that can provide a mocked `pipeline.Context`. For example, if we want to test controller logic without actual pods (i.e. using envtest) it's possible to provide a mocked context that returns an Infinispan client which always returns a successful status code whenever a call to the server is made.

My original intention was to add such tests in this PR, however it's already very large so I think that should come later. I also need to investigate golang mock libraries as there are many out there that generate the required mock code.

[the remark this answers is not in the document]
Comment (rigazilla) [reply pkg/reconcile/pipeline/infinispan/api.go:28]: Oh great! If this could ease a wider use of unit testing then these changes are more than welcome :)

**Comment by ryanemerson** [comment]
Thanks for reviewing @rigazilla

**Pull request merged**